File: RESTful service/UploadAudios.java

import java.io.File;
import java.io.IOException;
import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

// import commons-codec-<version>.jar, download from http://commons.apache.org/proper/commons-codec/download_codec.cgi
import org.apache.commons.codec.binary.Base64;

// import HttpClient,  download from http://hc.apache.org/downloads.cgi
/**
 *   
 *   commons-codec-1.1*.jar
 *   commons-logging-1.*.jar
 *   httpclient-4.*.jar
 *   httpcore-4.*.jar
 *   httpmime-4.*.jar
 * 
 * */
import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.HttpStatus;
import org.apache.http.client.config.RequestConfig;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.mime.MultipartEntityBuilder;
import org.apache.http.entity.mime.content.StringBody;
import org.apache.http.entity.ContentType;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;

public class UploadAudios {

	private String encodeBase64(byte[] bstr) {
		Base64 base64 = new Base64();
		return new String(base64.encode(bstr));
	}

	private String encryptByHMACSHA1(byte[] data, byte[] key) {
		try {
			SecretKeySpec signingKey = new SecretKeySpec(key, "HmacSHA1");
			Mac mac = Mac.getInstance("HmacSHA1");
			mac.init(signingKey);
			byte[] rawHmac = mac.doFinal(data);
			return encodeBase64(rawHmac);
		} catch (Exception e) {
			e.printStackTrace();
		}
		return "";
	}

	private String getUTCTimeSeconds() {
		Calendar cal = Calendar.getInstance();
		int zoneOffset = cal.get(Calendar.ZONE_OFFSET);
		int dstOffset = cal.get(Calendar.DST_OFFSET);
		cal.add(Calendar.MILLISECOND, -(zoneOffset + dstOffset));
		return cal.getTimeInMillis() / 1000 + "";
	}

	private String postHttp(String url, Map<String, Object> postParams,
			Map<String, String> headerParams, int timeout) {
		String result = null;

		if (postParams == null) {
			return result;
		}

		CloseableHttpClient httpClient = HttpClients.createDefault();
		try {
			HttpPost httpPost = new HttpPost(url);

			if (headerParams != null) {
				for (String key : headerParams.keySet()) {
					String value = headerParams.get(key);
					httpPost.addHeader(key, value);
				}
			}

			MultipartEntityBuilder mEntityBuilder = MultipartEntityBuilder
					.create();
			for (String key : postParams.keySet()) {
				Object value = postParams.get(key);
				if (value instanceof String || value instanceof Integer) {
					ContentType contentType = ContentType.create("text/plain", "UTF-8");
					StringBody stringBody = new StringBody(value+"", contentType);
					mEntityBuilder.addPart(key, stringBody);
				} else if (value instanceof File) {
					mEntityBuilder.addBinaryBody(key, (File) value);
				}
			}

			httpPost.setEntity(mEntityBuilder.build());

			RequestConfig requestConfig = RequestConfig.custom()
					.setConnectionRequestTimeout(timeout)
					.setConnectTimeout(timeout).setSocketTimeout(timeout)
					.build();
			httpPost.setConfig(requestConfig);

			HttpResponse response = httpClient.execute(httpPost);

			System.out.println(response.getStatusLine().getStatusCode());

			HttpEntity entity = response.getEntity();
			result = EntityUtils.toString(entity);
		} catch (Exception e) {
			e.printStackTrace();
		} finally {
			try {
				httpClient.close();
			} catch (IOException e) {
			}
		}
		return result;
	}

	public String upload(String audioId, String audioTitle, String dataPath,
			String dataType, String bucketName, String accessKey,
			String accessSecret, Map<String, String> userParams) {
		String result = null;
		String reqUrl = "https://api.acrcloud.com/v1/audios";
		String htttMethod = "POST";
		String httpAction = "/v1/audios";
		String signatureVersion = "1";
		String timestamp = this.getUTCTimeSeconds();

		File file = new File(dataPath);
		if (!file.exists()) {
			return null;
		}

		String sigStr = htttMethod + "\n" + httpAction + "\n" + accessKey
				+ "\n" + signatureVersion + "\n" + timestamp;
		String signature = encryptByHMACSHA1(sigStr.getBytes(),
				accessSecret.getBytes());

		Map<String, String> headerParams = new HashMap<String, String>();
		headerParams.put("access-key", accessKey);
		headerParams.put("signature-version", signatureVersion);
		headerParams.put("signature", signature);
		headerParams.put("timestamp", timestamp);

		Map<String, Object> postParams = new HashMap<String, Object>();
		postParams.put("title", audioTitle);
		postParams.put("audio_id", audioId);
		postParams.put("bucket_name", bucketName);
		postParams.put("data_type", dataType);
		postParams.put("audio_file", file);

		if (userParams != null) {
			int i = 0;
			for (String key : userParams.keySet()) {
				String value = userParams.get(key);
				postParams.put("custom_key[" + i + "]", key);
				postParams.put("custom_value[" + i + "]", value);
				i++;
			}
		}

		result = this.postHttp(reqUrl, postParams, headerParams, 8000);

		return result;
	}

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		String audioId = "XXX";
		String audioTitle = "xxx";
		String dataPath = "./a.mp3";
		String dataType = "audio"; // audio & fingerprint
		String bucketName = "<your bucket name>";
		String accessKey = "<your console access_key>";
		String accessSecret = "<your console access_secret>";

		Map<String, String> userParams = new HashMap<String, String>();
		userParams.put("<user-defined-key1>", "<user-defined-value1>");
		userParams.put("<user-defined-key2>", "<user-defined-value2>");

		UploadAudios ua = new UploadAudios();

		String result = ua.upload(audioId, audioTitle, dataPath, dataType,
				bucketName, accessKey, accessSecret, userParams);
		if (result == null) {
			System.out.println("upload error");
		}

		System.out.println(result);
	}

}

File: RESTful service/UploadAudios.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;


using System.Diagnostics;
using System.IO;

using System.Web;
using System.Security.Cryptography;
using System.Net;

namespace ACRCloudUploadFile
{
    class ACRCloudUploadFile
    {
        public static string PostHttp(string url, IDictionary<string, Object> headerParams, IDictionary<string, Object> postParams, int timeoutSecond)
        {
            string result = "";
            string BOUNDARYSTR = "acrcloud***copyright***2015***" + DateTime.Now.Ticks.ToString("x");
            string BOUNDARY = "--" + BOUNDARYSTR + "\r\n";
            var ENDBOUNDARY = Encoding.ASCII.GetBytes("--" + BOUNDARYSTR + "--\r\n\r\n");

            var stringKeyHeader = BOUNDARY +
                           "Content-Disposition: form-data; name=\"{0}\"" +
                           "\r\n\r\n{1}\r\n";
            var filePartHeader = BOUNDARY +
                            "Content-Disposition: form-data; name=\"{0}\"; filename=\"{1}\"\r\n" +
                            "Content-Type: application/octet-stream\r\n\r\n";

            var memStream = new MemoryStream();
            foreach (var item in postParams)
            {
                if (item.Value is string)
                {
                    string tmpStr = string.Format(stringKeyHeader, item.Key, item.Value);
                    byte[] tmpBytes = Encoding.UTF8.GetBytes(tmpStr);
                    memStream.Write(tmpBytes, 0, tmpBytes.Length);
                }
                else if (item.Value is byte[])
                {
                    var header = string.Format(filePartHeader, "audio_file", "audio_file");
                    var headerbytes = Encoding.UTF8.GetBytes(header);
                    memStream.Write(headerbytes, 0, headerbytes.Length);
                    byte[] sample = (byte[])item.Value;
                    memStream.Write(sample, 0, sample.Length);
                    memStream.Write(Encoding.UTF8.GetBytes("\r\n"), 0, 2);
                }
            }
            memStream.Write(ENDBOUNDARY, 0, ENDBOUNDARY.Length);

            HttpWebRequest request = null;
            HttpWebResponse response = null;
            Stream writer = null;
            StreamReader myReader = null;
            try
            {
                request = (HttpWebRequest)WebRequest.Create(url);
                request.Timeout = timeoutSecond * 1000;
                request.Method = "POST";
                foreach (var item in headerParams)
                {
                    if (item.Value is string)
                    {
                        request.Headers.Add(item.Key+":"+item.Value);
                    }
                }
                request.ContentType = "multipart/form-data; boundary=" + BOUNDARYSTR;

                memStream.Position = 0;
                byte[] tempBuffer = new byte[memStream.Length];
                memStream.Read(tempBuffer, 0, tempBuffer.Length);

                writer = request.GetRequestStream();
                writer.Write(tempBuffer, 0, tempBuffer.Length);
                writer.Flush();
                writer.Close();
                writer = null;

                response = (HttpWebResponse)request.GetResponse();
                myReader = new StreamReader(response.GetResponseStream(), Encoding.UTF8);
                result = myReader.ReadToEnd();
            }
            catch (WebException e)
            {
                Console.WriteLine("timeout:\n" + e.ToString());
            }
            catch (Exception e)
            {
                Console.WriteLine("other excption:" + e.ToString());
            }
            finally
            {
                if (memStream != null)
                {
                    memStream.Close();
                    memStream = null;
                }
                if (writer != null)
                {
                    writer.Close();
                    writer = null;
                }
                if (myReader != null)
                {
                    myReader.Close();
                    myReader = null;
                }
                if (request != null)
                {
                    request.Abort();
                    request = null;
                }
                if (response != null)
                {
                    response.Close();
                    response = null;
                }
            }

            return result;
        }

        public static string EncryptByHMACSHA1(string input, string key)
        {
            HMACSHA1 hmac = new HMACSHA1(System.Text.Encoding.UTF8.GetBytes(key));
            byte[] stringBytes = Encoding.UTF8.GetBytes(input);
            byte[] hashedValue = hmac.ComputeHash(stringBytes);
            return EncodeToBase64(hashedValue);
        }

        public static string EncodeToBase64(byte[] input)
        {
            string res = Convert.ToBase64String(input, 0, input.Length);
            return res;
        }

        public static string Upload(IDictionary<string, Object> audioParams, IDictionary<string, Object> userParams, int timeoutSecond)
        {
            string reqUrl = "https://api.acrcloud.com/v1/audios";
            string httpMethod = "POST";
            string httpAction = "/v1/audios";
            string signatureVersion = "1";
            string accessKey = (string)audioParams["access_key"];
            string accessSecret = (string)audioParams["access_secret"];
            string audioId = (string)audioParams["audio_id"];
            string audioTitle = (string)audioParams["audio_title"];
            string bucketName = (string)audioParams["bucket_name"];
            string dataType = (string)audioParams["data_type"];
            byte[] audioData = (byte[])audioParams["audio_data"];
            string timestamp = ((int)DateTime.UtcNow.Subtract(new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc)).TotalSeconds).ToString();

            string sigStr = httpMethod + "\n" + httpAction + "\n" + accessKey + "\n" + signatureVersion + "\n" + timestamp;
            string signature = EncryptByHMACSHA1(sigStr, accessSecret);

            Console.WriteLine(signature);

            var headerParams = new Dictionary<string, object>();
            headerParams.Add("access-key", accessKey);
            headerParams.Add("signature-version", signatureVersion);
            headerParams.Add("signature", signature);
            headerParams.Add("timestamp", timestamp);

            var postParams = new Dictionary<string, object>();
            postParams.Add("title", audioTitle);
            postParams.Add("audio_id", audioId);
            postParams.Add("bucket_name", bucketName);
            postParams.Add("data_type", dataType);
            postParams.Add("audio_file", audioData);

            if (userParams != null) {
			    int i = 0;
                foreach (var item in userParams)
                {
                    postParams.Add("custom_key[" + i + "]", item.Key);
                    postParams.Add("custom_value[" + i + "]", item.Value);
                    i++;
                }
		    }

            string res = PostHttp(reqUrl, headerParams, postParams, timeoutSecond);

            return res;
        }

        static void Main(string[] args)
        {
            string audioId = "XXX";
            string audioTitle = "xxx";
            string dataPath = "./a.mp3";
            string dataType = "audio"; // audio & fingerprint
            string bucketName = "<your bucket name>";
            string accessKey = "<your console access_key>";
            string accessSecret = "<your console access_secret>";

            var userParams = new Dictionary<string, object>();
            userParams.Add("<user-defined-key1>", "<user-defined-value1>");
            userParams.Add("<user-defined-key2>", "<user-defined-value2>");

            var audioParams = new Dictionary<string, object>();
            audioParams.Add("access_key", accessKey);
            audioParams.Add("access_secret", accessSecret);
            audioParams.Add("audio_id", audioId);
            audioParams.Add("audio_title", audioTitle);
            audioParams.Add("bucket_name", bucketName);
            audioParams.Add("data_type", dataType);
            

            using (FileStream fs = new FileStream(dataPath, FileMode.Open))
            {
                using (BinaryReader reader = new BinaryReader(fs))
                {
                    byte[] datas = reader.ReadBytes((int)fs.Length);
                    audioParams.Add("audio_data", datas);

                    // default timeout 10 seconds
                    string result = ACRCloudUploadFile.Upload(audioParams, userParams, 10);
                    Console.WriteLine(result);
                    Console.ReadLine();
                }
            }
        }
    }
}
